### DISCUSSION TITLE
Face recognition

### DISCUSSION THREAD

**Discussion by JoshuaSeidel** [blakeblackshear/frigate#15211] [Ask A Question]
Is there a place in the UI on 16 that you can identify faces (labeling) or teach frigate faces?

**Comment by NickM-27** [comment]
It is implemented in https://github.com/blakeblackshear/frigate/pull/15205

**Comment by JoshuaSeidel** [comment]
Awesome, looks like the faces page doesn't populate.

First there was an error that there was no clips/faces dir so I created it. And the errors went away but just a blank black page

**Reply by hawkeye217** [reply]
Just to be clear, these are completely unrefined features you're using that aren't meant for public consumption yet and are subject to massive amounts of change, so use at your own risk.

**Reply by JoshuaSeidel** [reply]
@hawkeye217 i know, I like alpha testing, I'm on a completely different instance then the instance that matters. I saw the features added and want to see how they work. I wasn't a fan of having 3-4 containers to get faces and the latency was too much for it to really be helpful, so it's exciting to see the frigate team implementing these natively

**Reply by NickM-27** [reply]
you will need to create the folders in there for the face names that you want to add for now, then there should be a plus button that lets you upload an image

**Reply by JoshuaSeidel** [reply]
Thanks @NickM-27 that worked. Will there eventually be an option to bulk upload photos to train?

**Reply by NickM-27** [reply]
Potentially, you can also just put the files directly into the folder, there is no requirement to use the UI

**Reply by JoshuaSeidel** [reply]
@NickM-27 so i put a bunch in the folder and they populated but look like the full image. when i upload them through the UI they only end up populating just the face, does this matter?

**Reply by NickM-27** [reply]
that is working correctly, it will automatically detect the face and then crop the image to it

**Reply by JoshuaSeidel** [reply]
@NickM-27 so should. i expect the ones i uploaded via SFTP to the server folder to crop in the UI at some point for the face? or is it showing the full picture it still takingin the face

**Reply by NickM-27** [reply]
if you upload manually you should crop before uploading

**Reply by JoshuaSeidel** [reply]
@NickM-27 is there a api endpoint i can just cycle through uploading images so frigate will do the cropping for me?

**Reply by NickM-27** [reply]
POST /faces/{face-name} where the file is named `file`

**Reply by JoshuaSeidel** [reply]
`curl -d @path/to/my-file.jpg http://192.168.*.*/faces/Joshua` ?

**Reply by JoshuaSeidel** [reply]
or 

`curl -F file=@path/to/my-file.jpg http://192.168.*.*/faces/Joshua` ?

**Reply by JoshuaSeidel** [reply]
@NickM-27 all tries got me this

`curl --request POST \
  --url http://192.168.9.144:5000/faces/Joshua \
  --header 'content-type: multipart/form-data' \
  --form file=@/Users/joshua.seidel/Desktop/faces/joshua/ffcf5f60-9c14-4bf7-85fc-80536b18059e.jpg`
  
```
  <html>
<head><title>405 Not Allowed</title></head>
<body><script>window.baseUrl="/";</script>
<center><h1>405 Not Allowed</h1></center>
<hr><center>nginx/1.25.3</center>
</body>
</html>
```

**Reply by NickM-27** [reply]
You'd need -X POST

**Comment by hadleyrich** [comment]
I was just playing around with this for fun. Looking cool!

No complaints as I'm aware it's pre-alpha, just trying to be of assistance. I noticed that the code that downloads the face models was removed in a commit:

https://github.com/blakeblackshear/frigate/commit/9c6e31a47a98c621907d068ec74fa34e47811a91#diff-97318136d136914ca20ef93a763955a8042b1c4f39133a850078a6da8a55248dL131

So if you try and upload a face through the UI currently it will fail here without a model:

https://github.com/blakeblackshear/frigate/blob/6475c551622c367ae9535f4272bcad73b5e7ec7a/frigate/embeddings/maintainer.py#L106

**Reply by NickM-27** [reply]
Oh yeah, I need to add that back in

**Comment by hadleyrich** [comment]
I was playing around with this some more this morning and noticed a couple of exceptions. Worked around them with this FYI;

```Diff
diff --git a/frigate/data_processing/real_time/face_processor.py b/frigate/data_processing/real_time/face_processor.py
index 5b0d6917..9be4ef5f 100644
--- a/frigate/data_processing/real_time/face_processor.py
+++ b/frigate/data_processing/real_time/face_processor.py
@@ -200,7 +200,11 @@ class FaceProcessor(RealTimeProcessorApi):
             y: int = max(raw_bbox[1], 0)
             w: int = raw_bbox[2]
             h: int = raw_bbox[3]
-            bbox = (x, y, x + w, y + h)
+            try:
+                bbox = (x, y, x + w, y + h)
+            except cv2.error:
+                logger.error("202 cv2.cvtColor failed.")
+
 
             if face is None or area(bbox) > area(face):
                 face = bbox
@@ -269,7 +273,10 @@ class FaceProcessor(RealTimeProcessorApi):
                 max(0, face_box[1]) : min(frame.shape[0], face_box[3]),
                 max(0, face_box[0]) : min(frame.shape[1], face_box[2]),
             ]
-            face_frame = cv2.cvtColor(face_frame, cv2.COLOR_RGB2BGR)
+            try:
+                face_frame = cv2.cvtColor(face_frame, cv2.COLOR_RGB2BGR)
+            except cv2.error:
+                logger.error("272 cv2.cvtColor failed.")
         else:
             # don't run for object without attributes
             if not obj_data.get("current_attributes"):
```
